src/storage/src/hummock/store/version.rs

```diff
@@ -456,32 +415,50 @@ impl HummockReadVersion {
     }
 
     pub fn add_merged_imm(&mut self, merged_imm: ImmutableMemtable) {
-        let staging_imm_count = self.staging.imm.len();
-        let merged_imm_ids = merged_imm.get_imm_ids();
-
-        #[cfg(debug_assertions)]
-        {
-            // check the suffix `merged_imm_ids.len()` imms in staging.imm are the same as
-            // `merged_imm_ids`
-            let diff = staging_imm_count - merged_imm_ids.len();
-            let mut count: usize = 0;
-            for (i, imm) in self.staging.imm.iter().skip(diff).enumerate() {
-                count += 1;
-                assert_eq!(
-                    imm.batch_id(),
-                    merged_imm_ids[i],
-                    "merged_imm_ids: {:?}",
-                    merged_imm_ids
-                );
+        assert!(merged_imm.get_imm_ids().iter().rev().is_sorted());
+        let min_imm_id = *merged_imm.get_imm_ids().last().expect("non-empty");
+
+        let back = self.staging.imm.back().expect("should not be empty");
+
+        // pop and save imms that are written earlier than the oldest imm if there is any
+        let earlier_imms = if back.batch_id() < min_imm_id {
+            let mut earlier_imms = VecDeque::with_capacity(self.staging.imm.len());
+            loop {
+                let batch_id = self
+                    .staging
+                    .imm
+                    .back()
+                    .expect("should not be empty")
+                    .batch_id();
+                match batch_id.cmp(&min_imm_id) {
+                    Ordering::Less => {
+                        let imm = self.staging.imm.pop_back().unwrap();
+                        earlier_imms.push_front(imm);
+                    }
+                    Ordering::Equal => {
+                        break;
+                    }
+                    Ordering::Greater => {
+                        unreachable!("must have break in equal")
+                    }
+                }
             }
-            assert_eq!(count, merged_imm_ids.len());
+            Some(earlier_imms)
+        } else {
+            assert_eq!(back.batch_id(), min_imm_id);
+            None
+        };
+
+        // iter from smaller imm and take the older imm at the back.
+        for imm_id in merged_imm.get_imm_ids().iter().rev() {
+            let imm = self.staging.imm.pop_back().expect("should exist");
+            assert_eq!(imm.batch_id(), *imm_id);
```

Comment:
let's log more useful information including staging imm ids, merged_imm ids, table_id and instance id here.